src/cromshell/cost/command.py

```diff
@@ -0,0 +1,302 @@
+import logging
+
+import click
+import cromshell.utilities.command_setup_utils as command_setup_utils
+import cromshell.utilities.workflow_id_utils as workflow_id_utils
+
+from datetime import datetime, timezone, timedelta
+from tempfile import NamedTemporaryFile
+from tabulate import tabulate
+
+LOGGER = logging.getLogger(__name__)
+
+
+@click.command(name="cost")
+@click.argument("workflow_id", required=True)
+@click.option(
+    "-d",
+    "--detailed",
+    is_flag=True,
+    default=False,
+    help="Get the cost for a workflow at the task level",
+)
+@click.pass_obj
+def main(config, workflow_id: str or int, detailed: bool):
+    """
+    Get the cost for a workflow."
+    Only works for workflows that completed more than 8 hours ago on GCS."
+    Requires the 'gcp_bq_cost_table.config' configuration file to exist and contain"
+    the big query cost table for your organization."
+
+    """
+
+    LOGGER.info("cost")
+
+    resolved_workflow_id = command_setup_utils.resolve_workflow_id_and_server(
+        workflow_id=workflow_id,
+        cromshell_config=config,
+    )
+
+    # check if workflow id exists
+    if not workflow_id_utils.workflow_id_exists(
+        workflow_id=resolved_workflow_id, submission_file=config.submission_file_path
+    ):
+        LOGGER.error("Could not find workflow id %s in submission file.", resolved_workflow_id)
+        raise ValueError(
+            f"Could not find workflow id {resolved_workflow_id} in submission file."
+        )
+
+    # check that bq setting in json config
+    check_cost_table_is_configured(config_options=config.cromshell_config_options)
+
+    # Get time workflow finished using metadata command (error if not finished)
+    LOGGER.info("Retrieving workflow metadata")
+    workflow_metadata = get_metadata(config=config)
+    start_time, end_time = get_submission_start_end_time(workflow_metadata=workflow_metadata)
+
+    LOGGER.info("Checking workflow completed and finished past 24hrs")
+    checks_before_query(end_time=end_time, workflow_id=resolved_workflow_id)
+
+    LOGGER.info("Querying BQ")
+    # Create start and end time for query, plus/minus a day from start and finish
+    query_start_time = str(datetime.strptime(start_time, "%Y-%m-%dT%H:%M:%S.%f%z") - timedelta(days=1))
+    query_end_time = str(datetime.strptime(end_time, "%Y-%m-%dT%H:%M:%S.%f%z") + timedelta(days=1))
+
+    query_results = query_bigquery(
+        workflow_id=resolved_workflow_id,
+        bq_database=config.cromshell_config_options["bq_cost_table"],
+        start_date=query_start_time,
+        end_date=query_end_time,
+        detailed=detailed
+    )
+
+    temp_query_result_csv = NamedTemporaryFile()
```

Comment:
Why are we writing all these temp files?  Can't we just do this processing in memory?

Reply:
updated